src/ferui-components/tree-view/tree-view-component.ts

```diff
@@ -0,0 +1,399 @@
+      }
+      .fui-virtual-scroller {
+        height: 100%;
+        width: 100%;
+      }
+      :host(.show-border) #fui-tree-view-component {
+        border: #e5eaee 1px solid;
+        padding: 0;
+      }
+    `,
+  ],
+})
+export class FuiTreeViewComponent<T> implements OnInit, OnDestroy {
+  @Input() treeNodeData: TreeNodeData<T>;
+
+  @Input() dataRetriever: TreeNodeDataRetriever<T> | PagedTreeNodeDataRetriever<T>;
+
+  @Input() config: TreeViewConfiguration;
+
+  @Input() loading: boolean = false;
+
+  @Input() error: boolean = false;
+
+  @HostBinding('class.show-border') border: boolean;
+
+  @ViewChild(FuiVirtualScrollerComponent) vs: FuiVirtualScrollerComponent;
+
+  @ViewChildren(FuiTreeNodeComponent) children: QueryList<FuiTreeNodeComponent<T>>;
+
+  @Output() onNodeEvent: EventEmitter<TreeViewEvent<any>> = new EventEmitter<TreeViewEvent<any>>();
+
+  treeViewStyles: FuiTreeViewComponentStyles;
+  colorTheme: TreeViewColorTheme;
+  scrollViewArray: TreeNode<T>[];
+  hasBorders: boolean;
+  BUFFER_AMOUNT: number = 1000;
+  private rootNode: TreeNode<T>;
+  private nonRootArray: TreeNode<T>[];
+  private scrollObservable: Observable<any>;
+  private scrollSubscription: Subscription;
+  private indexChangeSubscription: Subscription;
+  private originalWidth: number;
+  private scrollPromise: boolean = false;
+  private SERVER_SIDE_COMPONENT: boolean = false;
+  private TREE_NODE_COMPONENT_HEIGHT: number = 34;
+  private DEFAULT_WIDTH: string = 'auto';
+  private DEFAULT_HEIGHT: string = '100%';
+  private DEFAULT_COLOR_THEME: TreeViewColorTheme = 'WHITE';
+  // Default Throttle time fro scroll event
+  private THROTTLE_TIME: number = 500;
+
+  constructor(private el: ElementRef) {}
+
+  /**
+   * Initializes the Tree View component and all properties needed depending on inputs configuration
+   */
+  ngOnInit(): void {
+    this.SERVER_SIDE_COMPONENT = this.dataRetriever.hasOwnProperty('getPagedChildNodeData');
+    this.treeViewStyles = {
+      width: this.config.width ? this.config.width : this.DEFAULT_WIDTH,
+      height: this.config.height ? this.config.height : this.DEFAULT_HEIGHT,
+    };
+    this.indexChangeSubscription = this.vs.vsChange.subscribe(() => this.calculateScrollerWidth());
+    this.colorTheme = this.config.colorVariation ? this.config.colorVariation : this.DEFAULT_COLOR_THEME;
+    this.border = this.hasBorders = this.config.hasBorders ? this.config.hasBorders : false;
+    if (this.SERVER_SIDE_COMPONENT) {
+      this.BUFFER_AMOUNT = this.config.bufferAmount || this.BUFFER_AMOUNT;
+      this.scrollObservable = fromEvent(this.vs.element.nativeElement, 'scroll');
+      const scrollThrottleTime = this.config.throttleTime || this.THROTTLE_TIME;
+      this.scrollSubscription = this.scrollObservable.pipe(throttleTime(scrollThrottleTime)).subscribe(() => {
+        this.handleScroll(this.vs.viewPortInfo.endIndex);
+      });
+    }
+    if (this.treeNodeData instanceof NonRootTreeNode) {
```

Comment:
a NonRootTreeNode should be a regular TreeNode. Because as a developper, we don't always know if our data would be a NonRootTreeNode or a regular TreeNode.